cmd/dist/fetch.go

```diff
@@ -0,0 +1,254 @@
+			var cancel func()
+			ctx, cancel = contextpkg.WithTimeout(ctx, timeout)
+			defer cancel()
+		}
+
+		var (
+			locator = context.Args().First()
+			args    = context.Args().Tail()
+		)
+
+		if locator == "" {
+			return fmt.Errorf("containerd: remote required")
+		}
+
+		if len(args) < 1 {
+			return fmt.Errorf("containerd: object required")
+		}
+
+		// TODO(stevvooe): Create "multi-fetch" mode that just takes a remote
+		// then receives object/hint lines on stdin, returning content as
+		// needed.
+
+		object := args[0]
+		hints := args[1:]
+
+		resolver, err := getResolver(ctx)
+		if err != nil {
+			return err
+		}
+
+		remote, err := resolver.Resolve(ctx, locator)
+		if err != nil {
+			return err
+		}
+
+		ctx = log.WithLogger(ctx, log.G(ctx).WithFields(
+			logrus.Fields{
+				"remote": locator,
+				"object": object,
+			}))
+
+		log.G(ctx).Infof("fetching")
+		rc, err := remote.Fetch(ctx, object, hints...)
+		if err != nil {
+			return err
+		}
+		defer rc.Close()
+
+		if _, err := io.Copy(os.Stdout, rc); err != nil {
+			return err
+		}
+
+		return nil
+	},
+}
+
+// NOTE(stevvooe): Most of the code below this point is prototype code to
+// demonstrate a very simplified docker.io fetcher. We have a lot of hard coded
+// values but we leave many of the details down to the fetcher, creating a lot
+// of room for ways to fetch content.
+
+// getResolver prepares the resolver from the environment and options.
+func getResolver(ctx contextpkg.Context) (remotes.Resolver, error) {
+	return remotes.ResolverFunc(func(ctx contextpkg.Context, locator string) (remotes.Remote, error) {
+		if !strings.HasPrefix(locator, "docker.io") {
+			return nil, errors.Errorf("unsupported locator: %q", locator)
+		}
+
+		var (
+			base = url.URL{
+				Scheme: "https",
+				Host:   "registry-1.docker.io",
+			}
+			prefix = strings.TrimPrefix(locator, "docker.io/")
+		)
+
+		token, err := getToken(ctx, "repository:"+prefix+":pull")
+		if err != nil {
+			return nil, err
+		}
+
+		return remotes.RemoteFunc(func(ctx contextpkg.Context, object string, hints ...string) (io.ReadCloser, error) {
+			ctx = log.WithLogger(ctx, log.G(ctx).WithFields(
+				logrus.Fields{
+					"prefix": prefix, // or repo?
+					"base":   base.String(),
+					"hints":  hints,
+				},
+			))
+
+			paths, err := getV2URLPaths(prefix, object, hints...)
+			if err != nil {
+				return nil, err
+			}
+
+			for _, path := range paths {
+				base.Path = path
+				url := base.String()
+				log.G(ctx).WithField("url", url).Debug("fetch content")
+
+				req, err := http.NewRequest(http.MethodGet, url, nil)
+				if err != nil {
+					return nil, err
+				}
+
+				req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+				for _, mediatype := range remotes.HintValues("mediatype", hints...) {
+					req.Header.Set("Accept", mediatype)
+				}
+
+				resp, err := ctxhttp.Do(ctx, http.DefaultClient, req)
+				if err != nil {
+					return nil, err
+				}
+
+				if resp.StatusCode > 299 {
```

Comment:
@samuelkarp I don't think there is any reason to refuse a possible correct response with a incorrect but successful status code. Writing this code makes status code migrations much easier in practice. Only in the case that the API says "202 does this and 204 does this" do we want to handle specific status codes.

100s are unrealistic to see here, as most are going to be handled by the underlying http package.

Either way, this code is just POC code. There is lot that needs to be added in the future, so nitting about this is a waste of time.